File: klodder/Classes/FileArchive.cpp
#include "Exception.h"
#include "FileArchive.h"
#include "StreamReader.h"
#include "StreamWriter.h"

#define TYPE_FILE 1

FileArchive::~FileArchive()
{
	Clear();
}

void FileArchive::Clear()
{
	for (size_t i = 0; i < mFileList.size(); ++i)
		delete mFileList[i];
	
	mFileList.clear();
}

void FileArchive::Add(const char* name, Stream* stream, int length)
{
	if (length < 0)
		length = stream->Length_get();
	
	stream->Seek(0, SeekMode_Begin);
	
	FileArchiveMember* member = new FileArchiveMember();

	member->mFileName = name;
	member->mData.StreamFrom(stream, length);
	
	mFileList.push_back(member);
}

void FileArchive::Add(const char* name, Stream* stream)
{
	Add(name, stream, -1);
}

void FileArchive::Load(Stream* stream)
{
	StreamReader reader(stream, false);
	
	while (!stream->EOF_get())
	{
		uint32_t type = reader.ReadUInt32();
		
		switch (type)
		{
			case TYPE_FILE:
			{
				uint32_t version = reader.ReadUInt32();
				
				switch (version)
				{
					case 1:
					{
						FileArchiveMember* member = new FileArchiveMember();
						
						// read filename
						
						member->mFileName = reader.ReadText_Binary();
						
						// read data
						
						uint32_t length = reader.ReadUInt32();
						
						member->mData.StreamFrom(stream, length);
						
						mFileList.push_back(member);

						LOG_DBG("read: %s, length=%lu", member->mFileName.c_str(), length);
						
						break;
					}
						
					default:
						throw ExceptionVA("unknown archive member version: %d", (int)version);
				}
				
				break;
			}
				
			default:
				throw ExceptionVA("unknown archive element type: %d", (int)type);
		}
	}
}

/*void FileArchive::Save(Stream* stream)
{
	SaveBegin(stream);
	
	for (size_t i = 0; i < mFileList.size(); ++i)
	{
		FileArchiveMember* member = mFileList[i];
		
		member->mData.Seek(0, SeekMode_Begin);
		
		SaveAdd(stream, member->mFileName.c_str(), &member->mData, member->mLength);
	}
	
	SaveEnd(stream);
}*/

void FileArchive::SaveBegin(Stream* stream)
{
	// nop
}

void FileArchive::SaveAdd(Stream* stream, const char* name, Stream* dataStream, int dataLength)
{
	StreamWriter writer(stream, false);
	
	int version = 1;
	
	switch (version)
	{
		case 1:
		{
			// write element type
			
			writer.WriteUInt32(TYPE_FILE);
			
			// write archive member version
			
			writer.WriteUInt32(version);
			
			// write filename
			
			writer.WriteText_Binary(name);
			
			// write file data
			
			uint32_t length = dataLength >= 0 ? dataLength : (dataStream->Length_get() - dataStream->Position_get());
			
			writer.WriteUInt32(length);
			
			StreamExtensions::StreamTo(dataStream, stream, 4096, length);
		
			break;
		}
			
		default:
			throw ExceptionVA("unknown archive version: %d", (int)version);
	}
}

void FileArchive::SaveAdd(Stream* stream, const char* name, Stream* dataStream)
{
	SaveAdd(stream, name, dataStream, -1);
}

void FileArchive::SaveEnd(Stream* stream)
{
}

Stream* FileArchive::GetStream(std::string name)
{
	for (size_t i = 0; i < mFileList.size(); ++i)
		if (mFileList[i]->mFileName == name)
			return &mFileList[i]->mData;

	throw ExceptionVA("stream not found: %s", name.c_str());
}

File: klodder/KlodderHQ/FileArchive.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace KlodderHQ
{
    public static class BinaryReaderExtensions
    {
        public static string ReadText_Binary(this BinaryReader self)
        {
            UInt32 length = self.ReadUInt32();

            byte[] bytes = new byte[length];

            self.BaseStream.Read(bytes, 0, (int)length);

            return Encoding.ASCII.GetString(bytes);
        }
    }

    public class FileArchiveMember
    {
        public string FileName;
        public Stream Stream;
    }

    public class FileArchive
    {
        public const int TYPE_FILE = 1;

        public void Load(string fileName)
        {
            FileStream stream = new FileStream(fileName, FileMode.Open, FileAccess.Read);

            BinaryReader reader = new BinaryReader(stream);

            while (stream.Position != stream.Length)
            {
                UInt32 type = reader.ReadUInt32();

                switch (type)
                {
                    case TYPE_FILE:
                        {
                            UInt32 version = reader.ReadUInt32();

                            switch (version)
                            {
                                case 1:
                                    {
                                        FileArchiveMember member = new FileArchiveMember();

                                        // read filename

                                        member.FileName = reader.ReadText_Binary();

                                        // read data

                                        UInt32 length = reader.ReadUInt32();

                                        MemoryStream dataStream = new MemoryStream();

                                        byte[] bytes = new byte[length];

                                        stream.Read(bytes, 0, (int)length);

                                        dataStream.Write(bytes, 0, (int)length);
                                        dataStream.Seek(0, SeekOrigin.Begin);

                                        member.Stream = dataStream;

                                        MemberDict.Add(member.FileName, member);

                                        break;
                                    }

                                default:
                                    throw X.Exception("unknown archive member version: {0}", version);
                            }

                            break;
                        }

                    default:
                        throw X.Exception("unknown archive element type: {0}", type);
                }
            }
        }

        public Stream GetStream(string name)
        {
            FileArchiveMember member;

            if (!MemberDict.TryGetValue(name, out member))
                return null;
            else
                return member.Stream;
        }

        public Dictionary<string, FileArchiveMember> MemberDict = new Dictionary<string, FileArchiveMember>();
    }
}
